pdr_backend/lake/fetch_ohlcv.py

```diff
@@ -85,27 +83,52 @@ def safe_fetch_ohlcv_dydx(
     try:
         if exch != "dydx":
             return None
-        sinceIso = since.to_iso_timestr()
         headers = {"Accept": "application/json"}
         response = requests.get(
-            f"https://indexer.dydx.trade/v4/candles/perpetualMarkets/{symbol}"
-            + f"?resolution={timeframe}&fromISO={sinceIso}&limit={limit}",
-            headers=headers,
-            timeout=20,
-        )
-        response = requests.get(
-            "https://indexer.dydx.trade/v4/candles/perpetualMarkets/BTC-USD"
-            + "?resolution=5MINS&fromISO=2024-02-27T00:00:00.000Z&limit=1",
+            f"{BASE_URL_DYDX}/{symbol}?resolution={timeframe}"
+            f"&fromISO={since.to_iso_timestr()}&limit={limit}",
             headers=headers,
             timeout=20,
         )
         data = response.json()
-        return data
+
+        dydx_data = []
+        key_name = next(iter(data))  # Get the first key in the dict
+        items = data[key_name]
+
+        if key_name == "candles" and items:
+            for item in items:
+                dt = datetime.strptime(item["startedAt"], "%Y-%m-%dT%H:%M:%S.%fZ")
+                timestamp = int(dt.timestamp() * 1000)
+                ohlcv_tuple = (
+                    timestamp,
+                    float_or_none(item["open"]),
+                    float_or_none(item["high"]),
+                    float_or_none(item["low"]),
+                    float_or_none(item["close"]),
+                    float_or_none(item["baseTokenVolume"]),
+                )
+                dydx_data.append(ohlcv_tuple)
+
+            return dydx_data
+
+        if key_name == "errors" and items:
+            errors = items[0]
+            error_msg = tuple(errors.items())
+            dydx_data.append(error_msg)
+            return dydx_data
+
+        return None
+
     except Exception as e:
         logger.warning("exchange: %s", e)
         return None
 
 
+def float_or_none(x: str) -> Optional[float]:
+    return float(x) if x is not None else None
+
+
```

Comment:
This method needs a decorator `@enforce_types` above it.

Then you will see that it will fail sometime, because the argument specification here says that `x` can only be a `str`, and not a None. What it _should_ do is allow for either `str` or `None`. How you do that is `x: Union[str, None]`. And in fact, because None is one of the types, there's an equivalent shorthand: `x: Optional[str]`.

Finally the logic to return the value can be switched around, so that you can say `is none` rather than `is not None`. It's cleaner.

I'll put it all together for you:
```python
@enforce_types
def float_or_none(x: Optional[str]) -> Optional[float]:
    return None if x is None else float(x)
```